✨ Add Hybrid Mapping Extension  + Hybrid Synthesis

lsschmid opened the pull request:
## Description

Thir PR extends the Hybrid Mapper within QMAP. In particular in includes the following features:

1. Extended Hybrid Mapping
   - added support for Bridge Gates
   - added initial Mapping strategies
   - added support for more atom moves (passby and flying ancillas [limited for now])
   - general bugfixes and improvements
2. Hybrid Synthesis
    - a variant of the Hybrid Mapper which can evaluate multiple possible circuits, map them all end then chooses the best option to map (used for the ZX paper)
    - also includes PyBindings to use it from qiskit 
3. NaViz support
    - The routed circuit can now be exported to `.naviz` and `.namachine` which can then be fed into [NaViz](https://github.com/munich-quantum-toolkit/naviz) 
4. Added missing documentation on both mappers above.
5. Some general bugfixes and improvements

Fixes #466 

## Checklist:

- [x] The pull request only contains commits that are focused and relevant to this change.
- [x] I have added appropriate tests that cover the new/changed functionality.
- [x] I have updated the documentation to reflect these changes.
- [x] I have added entries to the changelog for any noteworthy additions, changes, fixes, or removals.
- [x] I have added migration instructions to the upgrade guide (if needed).
- [x] The changes follow the project's style guidelines and introduce no new warnings.
- [x] The changes are fully tested and pass the CI checks.
- [x] I have reviewed my own code changes.

lsschmid commented on src/hybridmap/HybridAnimation.cpp:
```diff
@@ -16,212 +16,173 @@
+                                     flyingAncillaIdxPlusOne * offset};
   }
-  return initString;
 }
```
yes, this is intentionally. The idea is to generalize this later to support many flying ancillas. This then might need some refactoring. For now this should be fine.

lsschmid commented on src/hybridmap/NeutralAtomArchitecture.cpp:
```diff
@@ -227,6 +263,63 @@ std::vector<CoordIndex> NeutralAtomArchitecture::getNN(CoordIndex idx) const {
+                     shuttlingSpeedFactor) +
+      "\n\trz: " + std::to_string(this->getGateTime("x")) +
+      "\n\try: " + std::to_string(this->getGateTime("x")) +
+      "\n\tcz: " + std::to_string(this->getGateTime("cz")) +
```
All archs should have either x or none, so this should not be a problem.

lsschmid commented on src/hybridmap/HybridNeutralAtomMapper.cpp:
```diff
@@ -1057,30 +1356,25 @@ CoordIndices NeutralAtomMapper::getBestMovePos(const CoordIndices& gateCoords) {
+        "No move position found (check if enough free coords are available)");
+  }
+  return finalBestPos.coords;
 }
```
For now, this should be fine, as we never cared too much about really tight architectures. In the long run falling back to SWAP would indeed make most sense.

lsschmid commented on bindings/hybrid_mapper/hybrid_mapper.cpp:
```diff
@@ -146,63 +161,158 @@
+                adjMatrix[i].emplace_back(symAdjMatrix(i, j));
+              }
+            }
+            return adjMatrix;
```
they are already in there!?

lsschmid commented on src/hybridmap/NeutralAtomScheduler.cpp:
```diff
@@ -30,31 +30,33 @@
   std::vector<std::deque<std::pair<qc::fp, qc::fp>>> rydbergBlockedQubitsTimes(
-      arch.getNpositions(), std::deque<std::pair<qc::fp, qc::fp>>());
+      numCoords);
   qc::fp aodLastBlockedTime = 0;
```
As mentioned elsewhere, this is planned for later.

burgholzer commented on python/mqt/qmap/hybrid_mapper.pyi:
```diff
@@ -90,12 +68,55 @@ class NeutralAtomHybridArchitecture:
+    def schedule(
+        self, verbose: bool = ..., create_animation_csv: bool = ..., shuttling_speed_factor: typing.SupportsFloat = ...
+    ) -> dict[str, float]: ...
+    def set_parameters(self, params: MapperParameters) -> None: ...
```
@lsschmid the rabit is right here. this seems slightly inconsistent

lsschmid replied:
They should all return qasm strings. I renamed them to make it more explicit.

burgholzer commented on src/hybridmap/HybridNeutralAtomMapper.cpp:
```diff
@@ -37,136 +43,200 @@
+                << hardwareQubits.getCoordIndex(hwQubit) << "\n";
+    }
+    std::cout << "\n";
+  }
```
Holds for the entire file here: could it be worth to consistenly use SPDLOG here instead of standard output?

lsschmid commented:
@burgholzer I incorporated the suggested changes and fixed some more minor typos as suggested by CodeRabbit. I also added an entry to the Upgrading and ChangeLog file.
